components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,402 @@
+  if (field.empty()) {
+    return false;
+  }
+  if (consume_delim && !ConsumeDelim(tokenizer)) {
+    return false;
+  }
+  return true;
+}
+
+bool FillOptionalStringField(StringTokenizer* tokenizer,
+                             const std::string& msg_token,
+                             absl::optional<std::string>& field) {
+  CHECK(tokenizer);
+  std::string field_str;
+  if (!FillStringField(tokenizer, tokenizer->token(), msg_token, false,
+                       field_str)) {
+    return false;
+  }
+  field = field_str;
+  return true;
+}
+
+bool FillOptionalResourcesField(StringTokenizer* tokenizer,
+                                absl::optional<std::vector<GURL>>& field) {
+  CHECK(tokenizer);
+  if (!CheckField(tokenizer->token(), kResourcesToken)) {
+    return false;
+  }
+  std::vector<GURL> urls;
+  while (tokenizer->GetNext()) {
+    if (tokenizer->token_is_delim()) {
+      continue;
+    }
+    if (!CheckField(tokenizer->token(), kResourcesSeperator)) {
+      return false;
+    }
+    auto url_str = tokenizer->token().substr(strlen(kResourcesSeperator));
+    GURL url(url_str);
+    if (!url.is_valid()) {
+      return false;
+    }
+    urls.push_back(url);
+  }
+  if (!urls.size()) {
+    return false;
+  }
+  field = urls;
+  return true;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-2.3
+bool IsUnreservedChar(char c) {
+  if (base::IsAsciiAlphaNumeric(c)) {
+    return true;
+  }
+  switch (c) {
+    case '-':
+    case '.':
+    case '_':
+    case '~':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986#section-2.2
+bool IsSubDelimChar(char c) {
+  switch (c) {
+    case '!':
+    case '$':
+    case '&':
+    case '\'':
+    case '(':
+    case ')':
+    case '*':
+    case '+':
+    case ',':
+    case ';':
+    case '=':
+      return true;
+  }
+  return false;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc3986/#section-3.3
+bool IsPChar(char c) {
+  if (IsUnreservedChar(c)) {
+    return true;
+  }
+  if (IsSubDelimChar(c)) {
+    return true;
+  }
+  switch (c) {
+    case ':':
+    case '[':
+    case ']':
+    case '@':
+      return true;
+  }
+  return false;
+}
+
+}  // namespace
+
+mojom::SIWEMessagePtr SIWEMessageParser::Parse(const std::string& message) {
+  state_ = State::kStart;
+  StringTokenizer tokenizer(message, "\n");
+  tokenizer.set_options(StringTokenizer::RETURN_DELIMS);
+  auto result = mojom::SIWEMessage::New();
+  bool seeking_optional_fields = false;
+  while (tokenizer.GetNext() || seeking_optional_fields) {
+    const std::string& token = tokenizer.token();
+    switch (state_) {
+      case State::kStart: {
+        std::string::size_type n = token.find(kStartingToken);
+        if (!n || n == std::string::npos) {
+          return {};
+        }
+        // missing line feed
+        if (token.substr(n) != kStartingToken) {
+          return {};
+        }
+        const std::string& old_origin_str = token.substr(0, n);
```

Comment:
nit can we rename this to domain if my presumptions around not checking domain == origin (from requesting tab) is correct. See other comment in `eth_provider_impl.c` for reference.

Reply:
no, we represent `${scheme}:// ${domain}` in origin, renaming to domain would be confusing and makes it looks like we are stripping info.

Reply:
Makes sense. I misread the code. After I looked at the test you referenced above I realized it. resolving this one.